compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -490,13 +490,97 @@ static bool isIndexVariable(Symbol* sym, CForLoop* loop) {
+static Symbol* extractLowerBoundFromLoop(CForLoop* loop) {
+  if(BlockStmt* bs = toBlockStmt(loop->initBlockGet())) {
+    for_alist(entry, bs->body) {
+      if(CallExpr *call = toCallExpr(entry)) {
+        if(call->isPrimitive(PRIM_ASSIGN)) {
+          if(SymExpr *symExpr = toSymExpr(call->get(2))) {
+            return symExpr->symbol();
+          }
+        }
+      }
+    }
+  }
+
+  INT_FATAL("Unexpected lower bound for loop identified for extraction as GPU kernel");
+  return nullptr;
+}
+
+static VarSymbol* generateNewVariable(BlockStmt* insertionPoint, const char* name, Type* type) {
+  VarSymbol *var = new VarSymbol(name, type);
+  var->defPoint = new DefExpr(var);
+  insertionPoint->insertAtTail(var->defPoint);
+  return var;
+}
+
+static VarSymbol* generateBlockSizeComputation(BlockStmt* gpuLaunchBlock, CForLoop* loop) {
+  Symbol* ub = extractUpperBoundFromLoop(loop);
+  Symbol* lb = extractLowerBoundFromLoop(loop);
+
+  VarSymbol *varBoundDelta = generateNewVariable(gpuLaunchBlock, "blockDelta", dtInt[INT_SIZE_DEFAULT]);
+  VarSymbol *varBlockSize = generateNewVariable(gpuLaunchBlock, "blockSize", dtInt[INT_SIZE_DEFAULT]);
+
+  CallExpr *c1 = new CallExpr(PRIM_ASSIGN, varBoundDelta, new CallExpr(PRIM_SUBTRACT, ub, lb));
+  gpuLaunchBlock->insertAtTail(c1);
+
+  CallExpr *c2 = new CallExpr(PRIM_ASSIGN, varBlockSize, new CallExpr(PRIM_ADD, varBoundDelta, new_IntSymbol(1)));
+  gpuLaunchBlock->insertAtTail(c2);
+
+  return varBlockSize;
+}
+
+static VarSymbol* generateAssignmentToPrimitive(
+  FnSymbol* fn, const char *varName, PrimitiveTag prim, Type *primReturnType) {
+
+  VarSymbol *var = generateNewVariable(fn->body, varName, primReturnType);
+  CallExpr *c1 = new CallExpr(PRIM_MOVE, var, new CallExpr(prim));
+  fn->insertAtTail(c1);
+  return var;
+}
+
+static VarSymbol* generateIndexComputation(FnSymbol* fn, Symbol* sym) {
+  VarSymbol *varBlockIdxX = generateAssignmentToPrimitive(fn, "blockIdxX", PRIM_GPU_BLOCKIDX_X, dtInt[INT_SIZE_32]);
+  VarSymbol *varBlockDimX = generateAssignmentToPrimitive(fn, "blockDimX", PRIM_GPU_BLOCKDIM_X, dtInt[INT_SIZE_32]);
+  VarSymbol *varThreadIdxX = generateAssignmentToPrimitive(fn, "threadIdxX", PRIM_GPU_THREADIDX_X, dtInt[INT_SIZE_32]);
+
+  VarSymbol *tempVar = generateNewVariable(fn->body, "t0", dtInt[INT_SIZE_32]);
+  CallExpr *c1 = new CallExpr(PRIM_MOVE, tempVar, new CallExpr(PRIM_MULT, varBlockIdxX, varBlockDimX));
+  fn->insertAtTail(c1);
+
+  VarSymbol *tempVar1 = generateNewVariable(fn->body, "t1", dtInt[INT_SIZE_32]);
+  CallExpr *c2 = new CallExpr(PRIM_MOVE, tempVar1, new CallExpr(PRIM_ADD, tempVar, varThreadIdxX));
+  fn->insertAtTail(c2);
+
+  return tempVar1;
+}
+
+static  CallExpr* generateGPUCall(CForLoop* loop,
+                                  BlockStmt* gpuLaunchBlock,
+                                  FnSymbol* kernel,
+                                  std::vector<VarSymbol*> actuals) {
+  VarSymbol *varBlockSize = generateBlockSizeComputation(gpuLaunchBlock, loop);
```

Comment:
I very much prefer this to be done at the top level (e.g. right before the call to this function) and `varBlockSize` passed to this function as an argument instead.

Reply:
Sure, I've factored that out.